nautobot/extras/tests/test_api.py

```diff
@@ -2631,6 +2661,192 @@ def test_populate_relationship_associations_on_site_create(self):
+                "slug in the request body's relationships[“platform-devices”][”destination”][”objects”] list data"
+            }
+        ]
+        self.assertHttpStatus(response, 400)
+        self.assertEqual(expected_error_json, response.json())
+
+        # Try creating a platform when all required data is present
+
+        create_data["relationships"] = {"platform-devices": {"destination": {"objects": [{"pk": str(device.pk)}]}}}
+        response = self.client.post(
+            reverse(get_route_for_model(Platform, "list", api=True)),
+            data=create_data,
+            format="json",
+            **self.header,
+        )
+        self.assertHttpStatus(response, 201)
+
+    def test_create_tenant_with_required_platform_source_o2o(self):
+        """
+        Test  that we can't create a tenant without specifying a platform for this required relationship
+        1. Try creating a tenant when no platforms exist
+        2. Try creating a tenant with no specified platform data
+        3. Try creating a tenant when all required data is present
+        """
+        self.user.is_superuser = True
+        self.user.save()
+
+        create_data = {
+            "name": "New Tenant",
+        }
+
+        # Try creating a tenant when no platform exists
+        response = self.client.post(
+            reverse(get_route_for_model(Tenant, "list", api=True)),
+            data=create_data,
+            format="json",
+            **self.header,
+        )
+        expected_error_json = [
+            {
+                "cr_platform-tenant__source": "Tenants require a platform, but no platforms exist yet. "
+                "Create a platform by posting to /api/dcim/platforms/"
+            }
+        ]
+        self.assertHttpStatus(response, 400)
+        self.assertEqual(expected_error_json, response.json())
+
+        # Try creating a tenant with no specified platform data
+        platform = Platform.objects.create(name="Platform 1")
+        response = self.client.post(
+            reverse(get_route_for_model(Tenant, "list", api=True)),
+            data=create_data,
+            format="json",
+            **self.header,
+        )
+        expected_error_json = [
+            {
+                "platform-tenant": "You must specify a platform id or slug in the request "
+                "body's relationships[“platform-tenant”][”source”][”objects”] list data"
+            }
+        ]
+        self.assertHttpStatus(response, 400)
+        self.assertEqual(expected_error_json, response.json())
+
+        # Try creating a tenant when all required data is present
+        create_data["relationships"] = {"platform-tenant": {"source": {"objects": [{"pk": str(platform.pk)}]}}}
+        response = self.client.post(
+            reverse(get_route_for_model(Tenant, "list", api=True)),
+            data=create_data,
+            format="json",
+            **self.header,
+        )
+        self.assertHttpStatus(response, 201)
+
 
```

Comment:
Also need some tests for the `update` case, including IMHO at least the following:

- object is created without the relationship, then the relationship is marked as required, object is updated without the relationship (fails)
- object is created without the relationship, then the relationship is marked as required, object is updated to include the relationship (succeeds)
- object is created with the required relationship, object is updated without specifying any relationship data (succeeds, relationship remains in place)
- object is created with the required relationship, object is updated to remove the relationship data (fails)